### ISSUE TITLE
nginx + systemd confinement broken recently

### ISSUE THREAD

**Issue by martinetd** [NixOS/nixpkgs#96840]
**Describe the bug**

Evaluating my system stopped working since the last channel update (Jul 17), because systemd confinement can no longer evaluate package list with de3780b937d2984f9b5e20d191f23be4f857b3aa ("nixos/nginx: move configuration testing script into reload command")

Evaluating it in the repl yields the following:
```
nix-repl> config.systemd.services.nginx.confinement.packages
[ «error: The option value `systemd.services.nginx.confinement.packages.[definition 1-entry 1]' in `/nix/var/nix/profiles/per-user/root/channels/nixos/nixos/modules/security/systemd-confinement.nix' is not of type `string or package'.» "/nix/store/571jywninsmn87akpzlwcp69nwwyccnx-nginx-1.16.1/bin/nginx -c '/nix/store/xps2d9q3438rjhj279zin6zfcd3paayp-nginx.conf' -p '/var/spool/nginx'" «derivation /nix/store/04qlr05g3866a1kp80aawx320br617cn-nginx-1.16.1.drv» «derivation /nix/store/5yrk2mqbxj5n1hxmkjl67rdvhbjk1lq7-coreutils-8.31.drv» ]
```

**To Reproduce**

http://cgit.notk.org/asmadeus/nixos-config.git/tree/profiles/nginx.nix?id=201ce97c31d0678f3c3235ecd294f03c73e23212 gives most of the config required to have the breakage but...

1. `git clone http://git.notk.org/asmadeus/nixos-config.git`
2. `cd nixos-config; git checkout 201ce97c31d0678f3c3235ecd294f03c73e23212; ln -s machines/jormungand/configuration.nix .`
3. `nix repl ./nixos.nix`
4. evaluate `config.systemd.services.nginx.confinement.packages`

**Expected behavior**
evaluates to something

**Additional informations**
I was able to work around the issue by forcing the value, e.g.
```
systemd.services.nginx.confinement.packages = lib.mkForce [ config.services.nginx.package pkgs.coreutils ];
```

So the problem is purely in the way the package list is built that doesn't parse properly complex commands? or intermediate variables I'm not sure... The machine I'm testing this on is way too slow to test comfortably, I'll play around later this week if required.

**Notify maintainers**
@aszlig (and I guess @flokli as this concerns nginx)

**Metadata**
```
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.45, NixOS, 20.03.2593.aaa66d8d887 (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.6`
 - channels(root): `"nixos-20.03.2862.925ae0dee63"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
(not sure how that works, please adjust if required)
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
 - config.systemd.services.nginx.confinement.packages
# a list of nixos modules affected by the problem
module:
 - services.nginx
 - systemd.services.nginx.confinement
```

**Comment by ju1m** [comment]
To be confirmed, but I guess that the case where `ExecReload=` is a list is not handled by `systemd-confinement.nix`:
https://github.com/NixOS/nixpkgs/blob/73cff3d2c27c61804498bc697fe43cdbb4678b0c/nixos/modules/security/systemd-confinement.nix#L138

**Comment by martinetd** [comment]
I confirmed that this is the case with @ju1m yesterday: forcing `systemd.services.nginx.serviceConfig.ExecReload` value to a single item fixes it, back to a list with trivial values breaks again.

So the construction of the package list is not resilient enough to list values - ju1m provided a fix for me to try, which looks good to me but I didn't take time to test yet:
```diff
diff --git a/nixos/modules/security/systemd-confinement.nix b/nixos/modules/security/systemd-confinement.nix
index 0a400f1d535..2927d424a8a 100644
--- a/nixos/modules/security/systemd-confinement.nix
+++ b/nixos/modules/security/systemd-confinement.nix
@@ -135,7 +135,7 @@ in {
           ];
           execPkgs = lib.concatMap (opt: let
             isSet = config.serviceConfig ? ${opt};
-          in lib.optional isSet config.serviceConfig.${opt}) execOpts;
+          in lib.flatten (lib.optional isSet config.serviceConfig.${opt})) execOpts;
           unitAttrs = toplevelConfig.systemd.units."${name}.service";
           allPkgs = lib.singleton (builtins.toJSON unitAttrs);
           unitPkgs = if fullUnit then allPkgs else execPkgs;
```

one of us will probably open a PR sooner or later

**Comment by flokli** [comment]
Has this been fixed in the meantime, or a PR filed?

I'd assume this will break peoples setups when using the `systemd-confinement` module - having multiple `ExecStarts` etc. seems pretty common.

**Comment by martinetd** [comment]
Thanks for the reminder, hadn't taken time -- doing this now.

FWIW I'm not sure how many people actually use confinement ? But might as well fix for 20.09 :)
